google_guest_agent/hostname/hostname_linux.go

```diff
@@ -0,0 +1,92 @@
+//  Copyright 2019 Google Inc. All Rights Reserved.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package hostname
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"os/exec"
+	"regexp"
+	"syscall"
+)
+
+const (
+	platformHostsFile = "/etc/hosts"
+	newline           = "\n"
+)
+
+func initPlatform(context.Context) {}
+
+var setHostname = func(hostname string) error {
+	if err := syscall.Sethostname([]byte(hostname)); err != nil {
+		return err
+	}
+	if _, err := exec.LookPath("nmcli"); err == nil {
+		if err := exec.Command("nmcli", "general", "hostname", hostname).Run(); err != nil {
+			return err
+		}
+	}
+	if _, err := exec.LookPath("systemctl"); err != nil {
+		o, err := exec.Command("systemctl").Output()
+		if err != nil {
+			return fmt.Errorf("error checking for rsyslog: %s", err)
+		}
+		if regexp.MustCompile(`rsyslog.service[^\n]*running`).Match(o) {
+			if err := exec.Command("systemctl", "--no-block", "restart", "rsyslog").Run(); err != nil {
+				return err
+			}
+		}
+	} else {
+		if err := exec.Command("pkill", "-HUP", "syslogd").Run(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Make the write as atomic as possible by creating a temp file, restoring
+// permissions & ownership, writing data, syncing, and then overwriting.
+func overwrite(dst string, contents []byte) error {
```

Comment:
is this doing the same as [SaferWriteFile()](https://github.com/GoogleCloudPlatform/guest-agent/blob/997df9e0b5d5c5df00df9072ea3f539be785d7c7/utils/file.go#L28)? if it doesn't fit all your needs we may adapt it to do so.

Reply:
Almost, SaferWriteFile does not handle user and group ownership. Adapted and updated.